src/shared/Network/Socket.cpp

```diff
@@ -0,0 +1,320 @@
+/*
+* This file is part of the CMaNGOS Project. See AUTHORS file for Copyright information
+*
+* This program is free software; you can redistribute it and/or modify
+* it under the terms of the GNU General Public License as published by
+* the Free Software Foundation; either version 2 of the License, or
+* (at your option) any later version.
+*
+* This program is distributed in the hope that it will be useful,
+* but WITHOUT ANY WARRANTY; without even the implied warranty of
+* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with this program; if not, write to the Free Software
+* Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
+*/
+
+#include <string>
+#include <memory>
+#include <vector>
+#include <functional>
+
+#include <boost/asio.hpp>
+#include <boost/date_time/posix_time/posix_time_types.hpp>
+
+#include "Socket.hpp"
+#include "Log.h"
+
+using namespace MaNGOS;
+
+Socket::Socket(boost::asio::io_service &service, std::function<void (Socket *)> closeHandler)
+    : m_socket(service), m_address("0.0.0.0"), m_outBufferFlushTimer(service),
+      m_closeHandler(closeHandler), m_writeState(WriteState::Idle), m_readState(ReadState::Idle) {}
+
+bool Socket::Open()
+{
+    try
+    {
+        const_cast<std::string &>(m_address) = m_socket.remote_endpoint().address().to_string();
+    }
+    catch (boost::system::error_code& error)
+    {
+        sLog.outError("Socket::Open() failed to get remote address.  Error: %s", error.message().c_str());
+        return false;
+    }
+
+    m_outBuffer.reset(new PacketBuffer);
+    m_secondaryOutBuffer.reset(new PacketBuffer);
+    m_inBuffer.reset(new PacketBuffer);
+
+    StartAsyncRead();
+
+    return true;
+}
+
+void Socket::Close()
+{
+    assert(!IsClosed());
+
+    m_socket.shutdown(boost::asio::ip::tcp::socket::shutdown_both);
+    m_socket.close();
```

Comment:
Would that result in a crash?  I thought about adding an atomic boolean for 'closed', but it seemed like if there were an issue like the one you mentioned (admittedly I hadn't considered cache consistency across multiple cores in a CPU), I'm not sure that it would crash.  Worst case, the queued ASIO operations would be aborted, causing an error, which would cause a close, which would do nothing since by then all cores should have caught up to the initial close.